flang/lib/Lower/ConvertType.cpp

```diff
@@ -248,8 +248,12 @@ struct TypeBuilderImpl {
     // links, the fir type is built based on the ultimate symbol. This relies
     // on the fact volatile and asynchronous are not reflected in fir types.
     const Fortran::semantics::Symbol &ultimate = symbol.GetUltimate();
-    if (Fortran::semantics::IsProcedurePointer(ultimate))
-      TODO(loc, "procedure pointers");
+
+    if (Fortran::semantics::IsProcedurePointer(ultimate)) {
+      Fortran::evaluate::ProcedureDesignator proc(ultimate);
+      return Fortran::lower::translateSignature(proc, converter);
```

Comment:
-- " I think we should use it everywhere and that it would probably be simpler to return the signature type wrapped in a fir.boxproc here."
Agreed. As a matter of fact, I need to make this return a `BoxProcType` to support procedure pointer function result.

-- "Using the "true" type in fir.boxproc for procedure pointers while it was not done for procedure dummies sounds OK to me: procedure pointers usually comes with a PROCEDURE interface while procedure dummies do not in old Fortran."
The problem is that `fir::store` requires the same memory type. If a procedure pointer as a "true" `BoxProcType`, it would requires the RHS (address of a procedure) to also have the "true" type. We currently use "empty" `BoxProcType` for getting the address of a procedure in `convertProcedureDesignatorToHLFIR`. Can we using "empty" type for now until we update both places?

Reply:
That is OK, but you could also solve the fir.store problem with a fir.convert before the store if needed (builder.createStoreWithConvert will deal with creating the fir.convert if and only if it is needed).

Reply:
Ok. I got it. I will try to setup the "true" type for procedure pointer and do `builder.createStoreWithConvert` at pointer assignment and argument passing.

Reply:
Also to add that it seems we actually need both `BoxProcType` and the `FunctionType` before box it in order to do `EmboxProcOp (FunctionType -> BoxProcType`.
So if the `genType` here returns `BoxProcType`, we need something like
```
if (auto boxTy = symTy.dyn_cast<fir::BoxProcType>())
      symTy = fir::unwrapRefType(boxTy.getEleTy());
```
to get the `FunctionType`.

Reply:
Good point. I think the solution for this may just be to add some fir::factory::genNullBoxProc(builder, loc, boxProcType) helper in include/flang/Optimizer/Builder/FIRBuilder.h (similar to fir::factory::createUnallocatedBox for allocatable/pointers) so that this is hidden/centralized.

A solution could also be to allow fir::ZeroOp to operate with a fir.boxproc type (and skip the fir.emboxproc for this case), but you would need to extend [BoxedProcedure.cpp](https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/CodeGen/BoxedProcedure.cpp) pass to rewrite this fir::ZeroOp usage too. But you can still define some genNullBoxProc helper anyway.

Reply:
-- "I think the solution for this may just be to add some fir::factory::genNullBoxProc(builder, loc, boxProcType) helper in include/flang/Optimizer/Builder/FIRBuilder.h (similar to fir::factory::createUnallocatedBox for allocatable/pointers) so that this is hidden/centralized."
Exactly. I currently (local) have a new `createNullEmboxProc` that is inspired by `createUnallocatedBox` in `MutableBox.h`. I will move it to FIRBuilder.h as it is a more reasonable place.

I will look into the other suggestion later.